ethereum-rpc/src/constructors/express/routes/api/v1/tokens.js

```diff
@@ -4,1236 +4,1293 @@ const { body, query, param, validationResult } = require("express-validator");
+        let { tokenAddress = "", id = "" } = { ...req.query };
+        const tokenContractAddress = prefixContractAddress(
+          tokenAddress.toString("hex")
+        );
+        if (!web3.utils.isAddress(tokenContractAddress)) {
+          return res.status(500).json({ error: "invalid contract address" });
+        }
+        const contractAbi = abi.APIS_ERC721;
+        const contract = await contractInstance(
+          contractAbi,
+          tokenContractAddress,
+          address
+        );
+        contract.methods
+          .getApproved(id)
+          .call()
+          .then((result) => {
+            res.status(200).json({ result });
+          });
+      } catch (e) {
+        console.error(e);
+        return res.status(500).json({ error: [e] });
+      }
     }
-    const { address = "*" } = req.params;
-    let {
-     tokenAddress = "",
-     toAddress = "",
-     tokenId = "",
-     privateKey = "",
-    } = { ...req.query };
-    const tokenContractAddress = prefixContractAddress(
-     tokenAddress.toString("hex")
-    );
-    const recieverAddress = prefixContractAddress(toAddress.toString("hex"));
-    if (!web3.utils.isAddress(tokenAddress)) {
-     return res.status(500).json({ error: "invalid contract address" });
+  );
+  router.get(
+    "/:address/collection",
+    [param("address").trim().isString(), query("tokenAddress").trim().isString(),
+    param("address").custom((value, { req }) => {
+      if (!web3.utils.isAddress(value)) {
+        throw new error("invalid address");
+      }
+      return true
+    }),
+    ],
+    async (req, res, next) => {
+      try {
+        const errors = validationResult(req);
+        if (!errors.isEmpty()) {
+          return res.status(400).json({ errors: errors.array() });
+        }
+        const { address = "*" } = req.params;
+        let { tokenAddress = "" } = { ...req.query };
+        const tokenContractAddress = prefixContractAddress(
+          tokenAddress.toString("hex")
+        );
+        if (!web3.utils.isAddress(tokenContractAddress)) {
+          return res.status(500).json({ error: "invalid contract address" });
+        }
+        const contractAbi = abi.APIS_ERC721;
+        const contract = await contractInstance(
+          contractAbi,
+          tokenContractAddress,
+          address
+        );
+        const balance = await contract.methods.balanceOf(address).call()
+        if (balance == 0){
+          return res.status(500).json({ error : "token balance zero"})
+        }
+        const index = Number(balance) - 1
+        const tokenId = await contract.methods.tokenOfOwnerByIndex(address, index).call()
+        const id = Number(tokenId)
+        const uri = await contract.methods.tokenURI(id).call()
+        const collection = []
+        for (i = 0; i < balance; i++) {
+          let tokenObtained = await contract.methods.tokenOfOwnerByIndex(address, i).call()
+          collection.push(tokenObtained)
+        }
```

Comment:
This should be done in a single asynchronous promise, or in pre-defined batches:

```javascript
const collection = Promise.all(
  Array.from(Array(balance).keys())
    .map(i => contract.methods.tokenOfOwnerByIndex(address, i).call())
)
```